src/lib/OpenEXR/ImfContextInit.cpp

```diff
@@ -0,0 +1,207 @@
+//
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) Contributors to the OpenEXR Project.
+//
+
+#include "ImfContextInit.h"
+#include "ImfIO.h"
+
+#include <cinttypes>
+#include <climits>
+#include <cstring>
+#include <mutex>
+
+OPENEXR_IMF_INTERNAL_NAMESPACE_SOURCE_ENTER
+
+struct istream_holder
+{
+    istream_holder (IStream* s) : _stream (s)
+    {
+        if (_stream) _cur_offset = _stream->tellg ();
+    }
+
+#if ILMTHREAD_THREADING_ENABLED
+    std::mutex _mx;
+#endif
+    uint64_t _cur_offset = 0;
+    IStream* _stream;
+};
+
+static int64_t
+istream_read (
+    exr_const_context_t         ctxt,
+    void*                       userdata,
+    void*                       buffer,
+    uint64_t                    sz,
+    uint64_t                    offset,
+    exr_stream_error_func_ptr_t error_cb)
+{
+    istream_holder* ih = static_cast<istream_holder*> (userdata);
+
+    if (sz > INT_MAX)
+    {
+        error_cb (
+            ctxt,
+            EXR_ERR_READ_IO,
+            "Stream interface request to read block too large");
+        return -1;
+    }
+
+#if ILMTHREAD_THREADING_ENABLED
+    std::lock_guard<std::mutex> lk{ih->_mx};
+#endif
+
+    if (offset != ih->_cur_offset)
+    {
+        ih->_stream->seekg (offset);
+        ih->_cur_offset = ih->_stream->tellg ();
+        if (offset != ih->_cur_offset)
+        {
+            error_cb (
+                ctxt,
+                EXR_ERR_READ_IO,
+                "Unable to seek to desired offset %" PRIu64,
+                offset);
+            return -1;
+        }
+    }
+
+    int64_t nread = ih->_cur_offset;
+    try
+    {
+        ih->_stream->read (
+            static_cast<char*> (buffer), static_cast<int> (sz));
+    }
+    catch (...)
+    {
+        // bah, there could be two reasons for this, one is a
+        // legitimate error, the other is a read past the end of file
+        // (i.e. the core library tries to read a 4k block when
+        // parsing the header), let's let the core deal with that and
+        // clear errors
+        ih->_stream->clear ();
+        //error_cb (
```

Comment:
should this be marked with a TODO

Reply:
yeah, kinda sorta - in the core, I try to read in pages, which may go past the end of the stream, which the current stream implementation doesn't really support well. But yes, should probably revisit the current ImfStdIO streams and boost their API a bit, so will definitely do that...